File: src/dotnet/Core/Application/Services/MetadataFilterBuilder.cs
using Rag.Candidates.Core.Application.DTOs;
using Rag.Candidates.Core.Domain.Entities;
using Rag.Candidates.Core.Domain.Enums;

namespace Rag.Candidates.Core.Application.Services;

public sealed class MetadataFilterBuilder
{
    private readonly string _candidateIdField;
    private readonly string _seniorityField;
    private readonly string _yearsExperienceField;
    private readonly string _skillNameField;
    private readonly string _typeField;
    private readonly string _typeSkill;
    private readonly Dictionary<SeniorityLevel, int> _seniorityOrder;

    public MetadataFilterBuilder(
        string candidateIdField,
        string seniorityField,
        string yearsExperienceField,
        string skillNameField,
        string typeField,
        string typeSkill)
    {
        _candidateIdField = candidateIdField;
        _seniorityField = seniorityField;
        _yearsExperienceField = yearsExperienceField;
        _skillNameField = skillNameField;
        _typeField = typeField;
        _typeSkill = typeSkill;
        _seniorityOrder = new Dictionary<SeniorityLevel, int>
        {
            [SeniorityLevel.Intern] = 0,
            [SeniorityLevel.Junior] = 1,
            [SeniorityLevel.Mid] = 2,
            [SeniorityLevel.Senior] = 3,
            [SeniorityLevel.Lead] = 4,
            [SeniorityLevel.Principal] = 5,
            [SeniorityLevel.Staff] = 6
        };
    }

    public List<Dictionary<string, object>> BuildCandidateFilters(ParsedQuery parsedQuery)
    {
        var filters = new List<Dictionary<string, object>>();

        if (parsedQuery.MinSeniorityLevel.HasValue)
        {
            var seniorityFilters = BuildSeniorityFilter(parsedQuery.MinSeniorityLevel.Value);
            filters.AddRange(seniorityFilters);
        }

        if (parsedQuery.MinYearsExperience.HasValue)
        {
            var yearsFilter = BuildYearsFilter(parsedQuery.MinYearsExperience.Value);
            filters.Add(yearsFilter);
        }

        return filters;
    }

    public Dictionary<string, object>? BuildTechnologyFilters(ParsedQuery parsedQuery)
    {
        if (parsedQuery.RequiredTechnologies.Length == 0)
            return null;

        return new Dictionary<string, object>
        {
            ["$and"] = new object[]
            {
                new Dictionary<string, object> { [_typeField] = _typeSkill },
                new Dictionary<string, object> 
                { 
                    [_skillNameField] = new Dictionary<string, object> 
                    { 
                        ["$in"] = parsedQuery.RequiredTechnologies 
                    } 
                }
            }
        };
    }

    public List<AggregatedCandidate> FilterAggregatedCandidates(
        List<AggregatedCandidate> candidates,
        ParsedQuery parsedQuery)
    {
        var filtered = new List<AggregatedCandidate>();

        foreach (var candidate in candidates)
        {
            if (MatchesFilters(candidate, parsedQuery))
            {
                filtered.Add(candidate);
            }
        }

        return filtered;
    }

    private bool MatchesFilters(AggregatedCandidate candidate, ParsedQuery parsedQuery)
    {
        if (parsedQuery.MinSeniorityLevel.HasValue)
        {
            var candidateSeniorityStr = candidate.Metadata.TryGetValue(_seniorityField, out var seniorityObj)
                ? seniorityObj?.ToString()
                : null;

            if (!MeetsSeniorityRequirement(candidateSeniorityStr, parsedQuery.MinSeniorityLevel.Value))
            {
                return false;
            }
        }

        if (parsedQuery.MinYearsExperience.HasValue)
        {
            var candidateYears = candidate.Metadata.TryGetValue(_yearsExperienceField, out var yearsObj)
                ? Convert.ToInt32(yearsObj)
                : 0;

            if (candidateYears < parsedQuery.MinYearsExperience.Value)
            {
                return false;
            }
        }

        return true;
    }

    private bool MeetsSeniorityRequirement(string? candidateSeniorityStr, SeniorityLevel requiredSeniority)
    {
        if (string.IsNullOrWhiteSpace(candidateSeniorityStr))
            return false;

        var candidateSeniority = ParseSeniority(candidateSeniorityStr);
        if (!candidateSeniority.HasValue)
            return false;

        var requiredLevel = _seniorityOrder.TryGetValue(requiredSeniority, out var reqLevel) ? reqLevel : 0;
        var candidateLevel = _seniorityOrder.TryGetValue(candidateSeniority.Value, out var candLevel) ? candLevel : 0;

        return candidateLevel >= requiredLevel;
    }

    private SeniorityLevel? ParseSeniority(string seniorityStr)
    {
        var seniorityMap = new Dictionary<string, SeniorityLevel>
        {
            ["Intern"] = SeniorityLevel.Intern,
            ["Junior"] = SeniorityLevel.Junior,
            ["Mid"] = SeniorityLevel.Mid,
            ["Senior"] = SeniorityLevel.Senior,
            ["Lead"] = SeniorityLevel.Lead,
            ["Principal"] = SeniorityLevel.Principal,
            ["Staff"] = SeniorityLevel.Staff
        };

        return seniorityMap.TryGetValue(seniorityStr, out var level) ? level : null;
    }

    private List<Dictionary<string, object>> BuildSeniorityFilter(SeniorityLevel minSeniority)
    {
        var minLevel = _seniorityOrder.TryGetValue(minSeniority, out var level) ? level : 0;

        var validSeniorities = new List<string>();
        foreach (var (seniority, seniorityLevel) in _seniorityOrder)
        {
            if (seniorityLevel >= minLevel)
            {
                validSeniorities.Add(SeniorityLevelToString(seniority));
            }
        }

        if (validSeniorities.Count == 0)
            return new List<Dictionary<string, object>>();

        return new List<Dictionary<string, object>>
        {
            new Dictionary<string, object>
            {
                [_seniorityField] = new Dictionary<string, object>
                {
                    ["$in"] = validSeniorities
                }
            }
        };
    }

    private Dictionary<string, object> BuildYearsFilter(int minYears)
    {
        return new Dictionary<string, object>
        {
            [_yearsExperienceField] = new Dictionary<string, object>
            {
                ["$gte"] = minYears
            }
        };
    }

    private static string SeniorityLevelToString(SeniorityLevel level)
    {
        return level switch
        {
            SeniorityLevel.Intern => "Intern",
            SeniorityLevel.Junior => "Junior",
            SeniorityLevel.Mid => "Mid",
            SeniorityLevel.Senior => "Senior",
            SeniorityLevel.Lead => "Lead",
            SeniorityLevel.Principal => "Principal",
            SeniorityLevel.Staff => "Staff",
            _ => "Unknown"
        };
    }
}

File: src/python/core/application/services/metadata_filter_builder.py
from typing import List, Dict, Any, Optional
from ...domain.entities.parsed_query import ParsedQuery
from ...domain.enums.seniority_level import SeniorityLevel


class MetadataFilterBuilder:
    def __init__(
        self,
        candidate_id_field: str,
        seniority_field: str,
        years_experience_field: str,
        skill_name_field: str,
        type_field: str,
        type_skill: str
    ):
        self._candidate_id_field = candidate_id_field
        self._seniority_field = seniority_field
        self._years_experience_field = years_experience_field
        self._skill_name_field = skill_name_field
        self._type_field = type_field
        self._type_skill = type_skill
        self._seniority_order = {
            SeniorityLevel.INTERN: 0,
            SeniorityLevel.JUNIOR: 1,
            SeniorityLevel.MID: 2,
            SeniorityLevel.SENIOR: 3,
            SeniorityLevel.LEAD: 4,
            SeniorityLevel.PRINCIPAL: 5,
            SeniorityLevel.STAFF: 6
        }
    
    def build_candidate_filters(self, parsed_query: ParsedQuery) -> List[Dict[str, Any]]:
        filters = []
        
        if parsed_query.min_seniority_level:
            seniority_filters = self._build_seniority_filter(parsed_query.min_seniority_level)
            if seniority_filters:
                filters.extend(seniority_filters)
        
        if parsed_query.min_years_experience:
            years_filter = self._build_years_filter(parsed_query.min_years_experience)
            if years_filter:
                filters.append(years_filter)
        
        return filters
    
    def build_technology_filters(self, parsed_query: ParsedQuery) -> Optional[Dict[str, Any]]:
        if not parsed_query.required_technologies:
            return None
        
        return {
            "$and": [
                {self._type_field: self._type_skill},
                {self._skill_name_field: {"$in": parsed_query.required_technologies}}
            ]
        }
    
    def filter_aggregated_candidates(
        self,
        candidates: List,
        parsed_query: ParsedQuery
    ) -> List:
        filtered = []
        
        for candidate in candidates:
            if self._matches_filters(candidate, parsed_query):
                filtered.append(candidate)
        
        return filtered
    
    def _matches_filters(self, candidate, parsed_query: ParsedQuery) -> bool:
        if parsed_query.min_seniority_level:
            candidate_seniority_str = candidate.metadata.get(self._seniority_field)
            if not self._meets_seniority_requirement(
                candidate_seniority_str,
                parsed_query.min_seniority_level
            ):
                return False
        
        if parsed_query.min_years_experience:
            candidate_years = candidate.metadata.get(self._years_experience_field, 0)
            if not isinstance(candidate_years, (int, float)) or candidate_years < parsed_query.min_years_experience:
                return False
        
        return True
    
    def _meets_seniority_requirement(
        self,
        candidate_seniority_str: Optional[str],
        required_seniority: SeniorityLevel
    ) -> bool:
        if not candidate_seniority_str:
            return False
        
        candidate_seniority = self._parse_seniority(candidate_seniority_str)
        if not candidate_seniority:
            return False
        
        required_level = self._seniority_order.get(required_seniority, 0)
        candidate_level = self._seniority_order.get(candidate_seniority, 0)
        
        return candidate_level >= required_level
    
    def _parse_seniority(self, seniority_str: str) -> Optional[SeniorityLevel]:
        seniority_map = {
            "Intern": SeniorityLevel.INTERN,
            "Junior": SeniorityLevel.JUNIOR,
            "Mid": SeniorityLevel.MID,
            "Senior": SeniorityLevel.SENIOR,
            "Lead": SeniorityLevel.LEAD,
            "Principal": SeniorityLevel.PRINCIPAL,
            "Staff": SeniorityLevel.STAFF
        }
        return seniority_map.get(seniority_str)
    
    def _build_seniority_filter(self, min_seniority: SeniorityLevel) -> List[Dict[str, Any]]:
        min_level = self._seniority_order.get(min_seniority, 0)
        
        valid_seniorities = []
        for seniority, level in self._seniority_order.items():
            if level >= min_level:
                valid_seniorities.append(seniority.value)
        
        if not valid_seniorities:
            return []
        
        return [{self._seniority_field: {"$in": valid_seniorities}}]
    
    def _build_years_filter(self, min_years: int) -> Dict[str, Any]:
        return {self._years_experience_field: {"$gte": min_years}}
